### PR TITLE
many: add CLI support for getting and setting aspects

### PR DESCRIPTION
**Pull request by MiguelPires**
Adds a CLI for getting and setting data through aspects. Doesn't include support for "unset". The first commit is a refactor that changes where the result of a get is namespaced just to make it easier to combine results at the API (it's mostly test changes).

### CODE REVIEW

File: daemon/api_aspects.go [hunk 1]
```diff
@@ -115,6 +116,7 @@ func setAspect(c *Command, r *http.Request, _ *auth.UserState) Response {
 	// NOTE: could be sync but this is closer to the final version and the conf API
 	summary := fmt.Sprintf("Set aspect %s/%s/%s", account, bundleName, aspect)
 	chg := newChange(st, "set-aspect", summary, nil, nil)
+	chg.SetStatus(state.DoneStatus)
```
Comment (pedronis) [inline daemon/api_aspects.go:119]: do we need a test for this change?

**Review by pedronis — DISMISSED** [review]
thank you, did a pass, I'm not sure at all about introducing "aspect" subcommands if it's not the final direction

File: cmd/snap/main.go [hunk 2]
```diff
@@ -361,6 +380,18 @@ func Parser(cli *client.Client) *flags.Parser {
 	registerCommands(cli, parser, routineCommand, routineCommands, func(ci *cmdInfo) {
 		checkUnique(ci, "routine ")
 	})
+
+	// Add the aspect command
+	aspectCommand, err := parser.AddCommand("aspect", shortAspectHelp, longAspectHelp, &cmdAspect{})
```
[the remark this answers is not in the document]
Comment (MiguelPires) [reply cmd/snap/main.go:385]: Yes, it was just to separate it and also to make it easier to have some documentation without just putting it in the get/set docs and risk confusing people. But yes, I can use detect the acc/bundle/aspect syntax early on

File: cmd/snap/cmd_aspects.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+	for _, cmd := range cmds {
+		cmd.hidden = true
+	}
+}
+
+var shortAspectGetHelp = i18n.G("Get an aspect's data")
+var longAspectGetHelp = i18n.G(`
+The "aspect get" command returns the data stored at the locations specified by
+a list of dot-separated paths.
```
[the remark this answers is not in the document]
Comment (MiguelPires) [reply cmd/snap/cmd_aspects.go:77]: The intention was to make the fact that the path applies to a view, opposed to the actual storage, sort of invisible. That's the same reason why I called them paths and not requests, like we do in the aspects `aspects` to distinguish from the storage paths. I considered it an implementation detail that would be better to hide from the user. I'll change this in the get documentation and then hide it behind the flag when I add the experimental flag

File: cmd/snap/cmd_aspects.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+	if err != nil {
+		return err
+	}
+
+	raw, err := json.Marshal(value)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintln(Stdout, string(raw))
```
[the remark this answers is not in the document]
Comment (MiguelPires) [reply cmd/snap/cmd_aspects.go:119]: With the move we already get the same formatting that get has

File: cmd/snap/main.go [hunk 5]
```diff
@@ -361,6 +361,7 @@ func Parser(cli *client.Client) *flags.Parser {
 	registerCommands(cli, parser, routineCommand, routineCommands, func(ci *cmdInfo) {
 		checkUnique(ci, "routine ")
 	})
+
```
Comment (pedronis) [inline cmd/snap/main.go:364]: spurious?

File: cmd/snap/cmd_set.go [hunk 6]
```diff
@@ -126,3 +147,14 @@ func (x *cmdSet) Execute(args []string) error {
 
 	return nil
 }
+
+func validateAspectID(id string) error {
+	parts := strings.Split(id, "/")
+	for _, part := range parts {
+		if part == "" {
+			return fmt.Errorf(i18n.G("aspect identifier must conform to format: <account_id>/<bundle>/<aspect>"))
```
Comment (pedronis) [inline cmd/snap/cmd_set.go:155]: s/account_id/account-id/

File: cmd/snap/cmd_get.go [hunk 7]
```diff
@@ -242,7 +250,21 @@ func (x *cmdGet) Execute(args []string) error {
 	snapName := string(x.Positional.Snap)
 	confKeys := x.Positional.Keys
 
-	conf, err := x.client.Conf(snapName, confKeys)
+	var conf map[string]interface{}
+	var err error
+	// TODO: check that the experimental aspect feature is enabled
+	if len(strings.Split(string(x.Positional.Snap), "/")) == 3 {
```
Comment (pedronis) [inline cmd/snap/cmd_get.go:256]: this predicate could also be factored out into isAspectID  helper?

File: client/aspects.go [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+	endpoint := fmt.Sprintf("/v2/aspects/%s", aspectID)
+	_, err = c.doSync("GET", endpoint, query, nil, nil, &result)
+	if err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+func (c *Client) AspectSet(aspectID string, requests map[string]interface{}) (changeID string, err error) {
```
Comment (pedronis) [inline client/aspects.go:43]: maybe requests =>, requestValues or requestsToValues?

**Review by pedronis — COMMENTED** [review]
thank you for the changes, some small comments

File: cmd/snap/main.go (hunk 5 above)
Comment (MiguelPires) [reply cmd/snap/main.go:364]: Yes, leftover from the removal. Removed it now, thanks

File: cmd/snap/cmd_get_test.go [hunk 9]
```diff
@@ -224,3 +226,171 @@ func (s *SnapSuite) mockGetEmptyConfigServer(c *C) {
...
+
+	c.Check(s.Stdout(), Equals, `{
+	"abc": 1,
+	"xyz": false
+}
+`)
+	c.Check(s.Stderr(), Equals, "")
+}
+
+func (s *aspectsSetSuite) TestAspectGetInvalidAspectID(c *check.C) {
```
Comment (andrewphelpsj) [inline cmd/snap/cmd_get_test.go:392]: `check` is already `.` imported in this file.

File: client/aspects_test.go [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+func (cs *clientSuite) TestAspectSet(c *C) {
+	cs.status = 202
+	cs.rsp = `{"type": "async", "status-code": 202, "change": "123"}`
+
+	chgID, err := cs.cli.AspectSet("a/b/c", map[string]interface{}{"foo": "bar", "baz": json.Number("1")})
+	c.Check(err, IsNil)
+	c.Check(chgID, Equals, "123")
+	c.Check(cs.reqs, HasLen, 1)
```
Comment (andrewphelpsj) [inline client/aspects_test.go:49]: An `Assert` would make it so the subsequent checks don't panic.

**Review by andrewphelpsj — APPROVED** [review]
Looks good, thanks!

**Pull request merged**